pkg/migration/manager.go

```diff
@@ -0,0 +1,283 @@
+			return err
+		}
+	}
+
+	// TODO(irfansharif): We'll need to check RequiredNodes again to make sure
+	// that no new nodes were added to the cluster in the interim, and loop back
+	// around if so.
+	return nil
+}
+
+// MigrateTo runs the set of migrations required to upgrade the cluster version
+// to the provided target version.
+func (m *Manager) MigrateTo(ctx context.Context, targetV roachpb.Version) error {
+	// TODO(irfansharif): Should we inject every ctx here with specific labels
+	// for each migration, so they log distinctly? Do we need an AmbientContext?
+	ctx = logtags.AddTag(ctx, "migration-mgr", redact.Safe(targetV))
+
+	// TODO(irfansharif): We'll need to acquire a lease here and refresh it
+	// throughout during the migration to ensure mutual exclusion.
+
+	// TODO(irfansharif): We'll want to either write to a KV key to record the
+	// version up until which we've already migrated to, or consult the system
+	// table mentioned above. Perhaps it makes sense to consult any given
+	// `StoreClusterVersionKey`, since the manager here will want to push out
+	// cluster version bumps for vX before attempting to migrate into vX+1.
+
+	// TODO(irfansharif): After determining the last completed migration, if
+	// any, we'll be want to assemble the list of remaining migrations to step
+	// through to get to targetV. For now we've hard-coded this list.
+	var vs []roachpb.Version
+
+	// Hacks to introduce fake migrations on the fly.
+	//
+	// TODO(irfansharif): Remove these, and remove purge VersionNoopMigration
+	// from history.
+	Registry[targetV] = GenerateFakeMigrationFor(targetV)
+	vs = append(vs, targetV)
+
+	for _, version := range vs {
+		h := &Helper{Manager: m}
+		// Push out the version gate to every node in the cluster. Each node
+		// will persist the version, bump the local version gates, and then
+		// return. The migration associated with the specific version can assume
+		// that every node in the cluster has the corresponding version
+		// activated.
+		{
+			// First sanity check that we'll actually be able to perform the
+			// cluster version bump, cluster-wide.
+			//
+			// TODO(irfansharif): Do we want to worry about the case where we
+			// validate N nodes running the right binary versions, and the user
+			// adding another node that would no longer be able to? As written,
+			// we'll be bumping the version gates down below anyway.
+			op := &serverpb.ValidateTargetClusterVersionRequest{Version: &version}
+			if err := h.EveryNode(ctx, op); err != nil {
+				return err
+			}
+		}
+		{
+			// TODO(irfansharif): Should we rename this request type to
+			// something like "BumpClusterVersion" instead?
+			op := &serverpb.AckClusterVersionRequest{Version: &version}
+			if err := h.EveryNode(ctx, op); err != nil {
+				return err
+			}
+		}
+
+		// TODO(irfansharif): We'll want a testing override here to be able to
+		// stub out migrations as needed.
+
+		migration, ok := Registry[version]
```

Comment:
We also need to run a number of migrations, not just one, right?

Reply:
The code I was expecting here was something like this

```go
for _, v := range clusterversion.GetVersionsFromRange(initialVersion, targetVersion) {
   migration, ok := Registry[v]
   if !ok { continue }
   migration(ctx, &Helper{...})
}
```

Reply:
Also, we need "fence versions". We can't run the migration into, say, 21.1-15 before all nodes have acknowleged the fact that a migration is ongoing. So basically each version with work attached to it needs to have an "empty" version before it. Example:

Instead of
```
20.2-1 migrate lock table
20.2-2 migrate raft log
```

It would have to be
```
20.1-1 noop
20.1-2 migrate lock table
20.1-3 noop
20.1-4 migrate raft log
```

Otherwise, you could get into situations in which 20.1-4 is not active yet, but some nodes have already started migrating their lock tables. Maybe this is fine, but in particular the first migration in a version is critical. Without the noop, we might start migrating lock tables before bumping the clusterversion up from `20.1`, which means the operator can roll back to a 20.1 binary that doesn't understand the migrated lock table.
I think for now we can just add an init function that does some validation of the `Registry` to make sure that there's always a noop bump before any work bump. Later, if this is too annoying a papercut, we can change the way clusterversions are defined so that we assign the `unstable` component automatically, and so we'll just use `2*i+1` instead of `i` in the current enum.

Reply:
#57186.